core/src/main/java/org/mule/runtime/core/internal/routing/ChoiceRouter.java

```diff
@@ -6,51 +6,190 @@
+import org.mule.runtime.api.lifecycle.InitialisationException;
+import org.mule.runtime.api.lifecycle.Lifecycle;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.context.MuleContextAware;
 import org.mule.runtime.core.api.el.ExpressionManager;
 import org.mule.runtime.core.api.el.ExpressionManagerSession;
 import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.core.api.management.stats.RouterStatistics;
 import org.mule.runtime.core.api.processor.Processor;
-import org.mule.runtime.core.privileged.routing.RoutePathNotFoundException;
+import org.mule.runtime.core.privileged.processor.Router;
+import org.mule.runtime.core.privileged.routing.RouterStatisticsRecorder;
 
-import java.util.Optional;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import javax.inject.Inject;
 
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+
 /**
  * Routes the event to a single<code>MessageProcessor</code> using an expression to evaluate the event being processed and find
  * the first route that can be used.
  * <p>
  * If a default route has been configured and no match has been found, the default route will be used. Otherwise it continues the
  * execution through the next MP in the chain.
  */
-public class ChoiceRouter extends AbstractSelectiveRouter {
+public class ChoiceRouter extends AbstractComponent implements Router, RouterStatisticsRecorder, Lifecycle, MuleContextAware {
 
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private final List<ProcessorRoute> routes = new ArrayList<>();
+
+  private Processor defaultProcessor;
+  private RouterStatistics routerStatistics;
+  private MuleContext muleContext;
   private ExpressionManager expressionManager;
 
-  @Override
-  protected Optional<Processor> selectProcessor(CoreEvent event) {
-    try (ExpressionManagerSession emSession = expressionManager.openSession(getLocation(), event, NULL_BINDING_CONTEXT)) {
-      return getConditionalMessageProcessors().stream()
-          .filter(cmp -> emSession.evaluateBoolean(cmp.getExpression(), false, true))
-          .findFirst()
-          .map(cmp -> cmp.getMessageProcessor());
-    }
+  public ChoiceRouter() {
+    routerStatistics = new RouterStatistics(TYPE_OUTBOUND);
   }
 
   @Override
-  protected Processor getProcessorToRoute(CoreEvent event) throws RoutePathNotFoundException {
-    try {
-      return super.getProcessorToRoute(event);
-    } catch (RoutePathNotFoundException e) {
-      return event1 -> event1;
-    }
+  public void setMuleContext(MuleContext context) {
+    this.muleContext = context;
   }
 
   @Inject
   public void setExpressionManager(ExpressionManager expressionManager) {
     this.expressionManager = expressionManager;
   }
 
+  @Override
+  public void initialise() throws InitialisationException {
+    if (defaultProcessor == null) {
+      defaultProcessor = event -> event;
```

Comment:
`Function.identity()`

Reply:
For some reason it cannot coerce the Function.identity() into Processor.